central/apitoken/expiration/expiration.go

```diff
@@ -0,0 +1,186 @@
+		return
+	}
+
+	now := time.Now()
+	aboutToExpireDate := now.Add(expirationWindow)
+	staleNotificationDate := now.Add(-staleNotificationAge)
+	staleNotificationTimestamp := protoconv.ConvertTimeToTimestamp(staleNotificationDate)
+
+	scheduleCtx := sac.WithGlobalAccessScopeChecker(context.Background(),
+		sac.AllowFixedScopes(
+			sac.AccessModeScopeKeys(storage.Access_READ_ACCESS, storage.Access_READ_WRITE_ACCESS),
+			sac.ResourceScopeKeys(resources.Notifications),
+		),
+	)
+	var notificationSchedule *storage.NotificationSchedule
+	notificationSchedule, found, err := n.store.GetNotificationSchedule(scheduleCtx)
+	if err != nil {
+		log.Error(err)
+		return
+	}
+	if !found || notificationSchedule == nil {
+		notificationSchedule = &storage.NotificationSchedule{
+			LastRun: protoconv.ConvertTimeToTimestamp(staleNotificationDate.Add(-1 * time.Hour)),
+		}
+	}
+	if notificationSchedule.GetLastRun().Compare(staleNotificationTimestamp) >= 0 {
+		return
+	}
+	notificationSchedule.LastRun = protoconv.ConvertTimeToTimestamp(now)
+	err = n.store.UpsertNotificationSchedule(scheduleCtx, notificationSchedule)
+	if err != nil {
+		log.Error(errors.Wrap(err, "failed to update expired API token notification last run"))
+	}
+
+	expiringTokens, err := n.listItemsToNotify(now, aboutToExpireDate)
+	if err != nil {
+		log.Error(err)
+		return
+	}
+	err = n.notify(expiringTokens)
+	if err != nil {
+		log.Error(err)
+		return
+	}
+}
+
+func (n *expirationNotifierImpl) listItemsToNotify(now time.Time, expiresUntil time.Time) ([]*storage.TokenMetadata, error) {
+	formattedNow := now.Format(timestampLayout)
+	formattedExpiresUntil := expiresUntil.Format(timestampLayout)
+	// Search tokens that expire before expiresUntil, that have not expired yet,
+	// and that have not been notified since notifiedUntil.
+	// That is Expiration < expiresUntil and LastNotified < notifiedUntil.
+	queryNotExpired := search.NewQueryBuilder().
+		AddStrings(search.Expiration, fmt.Sprintf(">%s", formattedNow)).
+		ProtoQuery()
+	queryAboutToExpire := search.NewQueryBuilder().
+		AddStrings(search.Expiration, fmt.Sprintf("<%s", formattedExpiresUntil)).
+		ProtoQuery()
+	queryNotRevoked := search.NewQueryBuilder().
+		AddBools(search.Revoked, false).
+		ProtoQuery()
+	query := search.ConjunctionQuery(
+		queryNotExpired,
+		queryAboutToExpire,
+		queryNotRevoked,
+	)
+	response, err := n.store.SearchRawTokens(expirySearchCtx, query)
+	if err != nil {
+		return nil, err
+	}
+	return response, nil
+}
+
+func (n *expirationNotifierImpl) notify(items []*storage.TokenMetadata) error {
+	now := time.Now()
+	for _, token := range items {
+		expiration := protoconv.ConvertTimestampToTimeOrNow(token.GetExpiration())
+		ttl := expiration.Sub(now)
+		timeUnit := int(expirationSlice.Seconds())
+		ttlSeconds := int(ttl.Seconds())
+		sliceCount := ttlSeconds / timeUnit
+		if ttlSeconds%timeUnit != 0 {
+			sliceCount++
+		}
+		sliceDuration := "hours"
+		if sliceCount == 1 {
+			sliceDuration = "hour"
+		}
```

Comment:
How about factoring this out into. separate function, let it return the "log message" so we also actually test the log message generation? Currently, it's not tested at all but AFAICT this is the "essence" of this PR: providing an accurate log statement.